File: code/3.py
import numpy as np
import matplotlib.pyplot as plt


a0 = 0.1  # MPa
f0 = 2.0  # MHz
w0 = 2 * np.pi * f0
T = 0.5  # мкс
N = 16   # points
h = T / N  # 0.03125 microseconds
fs = 1 / h  # 32 MHz


l_indices = np.arange(N)
t_l = l_indices * h
p_l = 2 * a0 * np.sin(3 * w0 * t_l) + a0 * np.cos(5 * w0 * t_l)

p_T_discrete = np.zeros(N, dtype=complex)

half_N = N // 2
for n in range(half_N + 1):
    sum_val = 0j
    for l in range(N):
        exponent = -1j * 2 * np.pi * n * l / N
        sum_val += p_l[l] * np.exp(exponent)
    p_T_discrete[n] = sum_val / N


for n in range(half_N + 1, N):
    p_T_discrete[n] = np.conj(p_T_discrete[N - n])

n_extended = np.arange(-N, N + 1)
freqs_extended = n_extended * (fs / N)
p_T_extended = p_T_discrete[n_extended % N]

p_T_extended[np.abs(p_T_extended) < 1e-10] = 0

amplitude_ext = np.abs(p_T_extended)
phase_ext = np.angle(p_T_extended)
real_part_ext = np.real(p_T_extended)
imag_part_ext = np.imag(p_T_extended)


fig = plt.figure(figsize=(12, 10))

plt.subplot(2, 2, 1)
plt.stem(freqs_extended, amplitude_ext, basefmt=" ")
plt.title('Амплитудный спектр ДПФ')
plt.xlabel('Частота, МГц')
plt.ylabel('Амплитуда, МПа')
plt.axvline(x=fs/2, color='r', linestyle='--', alpha=0.5)
plt.axvline(x=-fs/2, color='r', linestyle='--', alpha=0.5)
plt.grid(True)

plt.subplot(2, 2, 2)
plt.stem(freqs_extended, phase_ext, basefmt=" ")
plt.title('Фазовый спектр ДПФ')
plt.xlabel('Частота, МГц')
plt.ylabel('Фаза, рад')
plt.grid(True)

plt.subplot(2, 2, 3)
plt.stem(freqs_extended, real_part_ext, basefmt=" ")
plt.title('Действительная часть ДПФ')
plt.xlabel('Частота, МГц')
plt.grid(True)

plt.subplot(2, 2, 4)
plt.stem(freqs_extended, imag_part_ext, basefmt=" ")
plt.title('Мнимая часть ДПФ')
plt.xlabel('Частота, МГц')
plt.grid(True)

plt.tight_layout()
plt.savefig('fig_3.png', dpi=300)

File: code/3.m
clear; clc; close all;

blue_color = [0.12, 0.47, 0.71];
red_color  = [0.84, 0.15, 0.16];
set(0, 'DefaultFigureColor', 'w');
set(0, 'DefaultAxesColor', 'w');
set(0, 'DefaultTextFontName', 'DejaVu Sans');
set(0, 'DefaultAxesFontName', 'DejaVu Sans');
set(0, 'DefaultAxesFontSize', 10);
set(0, 'DefaultTextFontSize', 12);
set(0, 'DefaultAxesLineWidth', 1.5);
set(0, 'DefaultLineLineWidth', 1.7);
set(0, 'DefaultTextFontWeight', 'normal');
set(0, 'DefaultAxesTitleFontWeight', 'normal'); 
set(0, 'DefaultAxesTickDir', 'in');
set(0, 'DefaultAxesTickLength', [0.015 0.025]);

set(0, 'DefaultAxesXGrid', 'on');
set(0, 'DefaultAxesYGrid', 'on');
set(0, 'DefaultAxesGridLineStyle', '-');
set(0, 'DefaultAxesGridColor', [0.6 0.6 0.6]);
set(0, 'DefaultAxesGridAlpha', 0.5);

a0 = 0.1;
f0 = 2.0;
w0 = 2 * pi * f0;
T = 0.5;
N = 16;
h = T / N;
fs = 1 / h;

l_indices = 0:1:N;
t_l = l_indices * h;
p_l = 2 * a0 * sin(3*w0 * t_l) + a0 * cos(5 * w0 * t_l);
p_T_discrete = zeros(N);

half_N = N / 2;
for n = 0 : half_N
    sum_val = 0;
    for l = 0:(N-1)
        exponent = -1i * 2 * pi * n * l / N;
        sum_val = sum_val + p_l(l+1) * exp(exponent);
    end
    p_T_discrete(n+1) = sum_val / N;
end

for n = half_N+1:(N-1)
    p_T_discrete(n) = conj(p_T_discrete((N - n) + 1));
end

n_extended = -N:1:N;
freqs_extended = n_extended .* (fs / N);
p_T_extended = p_T_discrete(mod(n_extended, N) + 1);
p_T_extended(abs(p_T_extended) < 1e-10) = 0;

ampliude_ext = abs(p_T_extended);
phase_ext = angle(p_T_extended);
real_part_ext = real(p_T_extended);
imag_part_ext = imag(p_T_extended);


figure('Position', [110, 55, 1600, 1300]);

axes('Position', [0.0628, 0.5508, 0.4298, 0.4129]); 
stem(freqs_extended, ampliude_ext, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color);
hold on;
y_bounds = ylim;
plot([fs/2, fs/2], y_bounds, '--r', 'LineWidth', 1.7, 'Color', red_color);
plot([-fs/2, -fs/2], y_bounds, '--r', 'LineWidth', 1.7, 'Color', red_color);
hold off;
title('\bf a) \rm  Амплитудный спектр ДПФ');
xlabel('Частота, МГц');
ylabel('Амплитуда, МПа');
grid on;

axes('Position', [0.5577, 0.5508, 0.4298, 0.4129]); 
stem(freqs_extended, phase_ext, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color)
title('\bf b) \rm  Фазовый спектр ДПФ');
xlabel('Частота, МГц');
ylabel('Фаза, рад');
grid on;

axes('Position', [0.0628, 0.0583, 0.4298, 0.4129]); 
stem(freqs_extended, real_part_ext, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color)
title('\bf c) \rm  Действительная часть ДПФ');
xlabel('Частота, МГц');
grid on;

axes('Position', [0.5577, 0.0583, 0.4298, 0.4129]); 
stem(freqs_extended, imag_part_ext, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color)
title('\bf d) \rm  Мнимая часть ДПФ');
xlabel('Частота, МГц');
grid on;

print(gcf, 'fig_3_m.png', '-dpng', '-r300');
